api4/user.go

```diff
@@ -1129,6 +1153,80 @@ func updateUser(c *Context, w http.ResponseWriter, r *http.Request) {
 	w.Write([]byte(ruser.ToJson()))
 }
 
+func isOauth(userService string) bool {
+	for _, service := range []string{
+		model.SERVICE_GOOGLE,
+		model.SERVICE_OFFICE365,
+		model.SERVICE_OPENID,
+		model.SERVICE_GITLAB,
+	} {
+		if service == userService {
+			return true
+		}
+	}
+	return false
+}
+
+func checkLoginProviderAttributes(
+	c *Context, ouser *model.User, patch *model.UserPatch, methodName string) {
+	errName := "api.user.patch_user.login_provider_attribute_set.app_error"
+	if methodName == "updateUser" {
+		errName = "api.user.update_user.login_provider_attribute_set.app_error"
+	}
+	const (
+		LDAP = model.USER_AUTH_SERVICE_LDAP
+		SAML = model.USER_AUTH_SERVICE_SAML
+	)
+	LS := &c.App.Config().LdapSettings
+	SS := &c.App.Config().SamlSettings
+	// If LDAP or SAML is used, and either the first name or the last name are already set,
+	// or OAUTH is used, then neither the first name nor the last name may be modified
+	if ((patch.FirstName != nil && *patch.FirstName != ouser.FirstName) ||
+		(patch.LastName != nil && *patch.LastName != ouser.LastName)) &&
+		((ouser.AuthService == LDAP && (*LS.FirstNameAttribute != "" || *LS.LastNameAttribute != "")) ||
+			(ouser.AuthService == SAML && (*SS.FirstNameAttribute != "" || *SS.LastNameAttribute != "")) ||
+			isOauth(ouser.AuthService)) {
+		c.Err = model.NewAppError(
+			methodName, errName,
+			map[string]interface{}{"Field": "full name"}, "", http.StatusConflict)
+		return
+	}
+	// If LDAP or SAML is used, and any of the following attributes have been specified
+	// in the settings, then their corresponding fields may not be changed
+	for _, attr := range []struct {
```

Comment:
Rather than doing this loop 6 times, every time. Create the array necessary for LDAP and/or SAML and loop through that array. Which means you won't loop at all for OAUTH or "".